### PR TITLE
Cherry-pick documentation for cost, statistics, CBO

### PR DESCRIPTION
**Pull request by aweisberg**
Cherry-pick of prestosql/presto#127

### CODE REVIEW

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+:doc:`/optimizer/statistics` provided by connectors to estimate
+the costs for different join orders and automatically pick the
+join order with the lowest computed costs.
+
+The join enumeration strategy is governed by the ``join_reordering_strategy``
+session property, with the ``optimizer.join-reordering-strategy``
+configuration property providing the default value.
+
+The valid values are:
+ * ``AUTOMATIC`` (default) - full automatic join enumeration enabled
```
Comment (arhimondr) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: It is not a "default" just yet, but will be soon: https://github.com/prestodb/presto/pull/12752

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+With cost based join distribution selection, Presto automatically chooses to
+use a partitioned or broadcast join. With cost based join enumeration, Presto
+automatically chooses which side is the probe and which is the build.
+
+The join distribution strategy is governed by the ``join_distribution_type ``
+session property, with the ``join-distribution-type`` configuration property
+providing the default value.
+
+The valid values are:
+ * ``AUTOMATIC`` (default) - join distribution type is determined automatically
```
Comment (arhimondr) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: ditto

File: presto-docs/src/main/sphinx/optimizer/statistics.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+   * **low value**: the smallest value in the column
+   * **high value**: the largest value in the column
+
+The set of statistics available for a particular query depends on the connector
+being used and can also vary by table or even by table layout. For example, the
+Hive connector does not currently provide statistics on data size.
+
+Table statistics can be displayed via the Presto SQL interface using the
+:doc:`/sql/show-stats` command. For the Hive connector, refer to the
+:ref:`Hive connector <hive_analyze>` documentation to learn how to update table
```
Comment (arhimondr) [inline presto-docs/src/main/sphinx/optimizer/statistics.rst]: We should add a few words here pointing to the automatic stats collection. But let's do this in a separate PR

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst (hunk 1 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: yeah, i would change this for now to mark ELIMINATE_CROSS_JOINS as the default and update the documentation in #12752

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+---------------------------
+
+Presto uses a hash based join algorithm. That implies that for each join
+operator a hash table must be created from one join input (called build side).
+The other input (probe side) is then iterated and for each row the hash table is
+queried to find matching rows.
+
+There are two types of join distributions:
+ * Partitioned: each node participating in the query builds a hash table
+   from only fraction of the data
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: nit: "a fraction"

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ * Broadcast: each node participating in the query builds a hash table
+   from all of the data (data is replicated to each node)
+
+Each type have their trade offs. Partitioned joins require redistributing both
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: grammar nit: Each type has its trade offs

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+tables using a hash of the join key. This can be slower (sometimes
+substantially) than broadcast joins, but allows much larger joins. In
+particular, broadcast joins will be faster if the build side is much smaller
+than the probe side. However, broadcast joins require that the tables on the
+build side of the join after filtering fit in memory on each node, whereas
+distributed joins only need to fit in distributed memory across all nodes.
+
+With cost based join distribution selection, Presto automatically chooses to
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: nit: chooses *whether* to use

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+use a partitioned or broadcast join. With cost based join enumeration, Presto
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: Decision of build and probe side is also done by cost-based join distribution.  Cost-based join-enumeration adds the decision of whether to reorder multiple inner joins (if you're joining 3 tables, which 2 get joined first)

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst (hunk 2 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: current default is "PARTITIONED"

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+ * ``AUTOMATIC`` (default) - join distribution type is determined automatically
+   for each join
+ * ``BROADCAST`` - broadcast join distribution is used for all joins
+ * ``PARTITIONED`` - partitioned join distribution is used for all join
+
+Connector Implementations
+-------------------------
+
+In order for the Presto optimizer to use the cost based strategies,
+the connection implementation must provide :doc:`statistics`.
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: connection -> connector

File: presto-docs/src/main/sphinx/optimizer/statistics.rst (hunk 3 above)
Comment (aweisberg) [reply presto-docs/src/main/sphinx/optimizer/statistics.rst]: This doesn't require action so resolving.

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst (hunk 1 above)
Comment (aweisberg) [reply presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: Changed to reflect correct default of ELIMINATE_CROSS_JOINS

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst (hunk 7 above)
Comment (aweisberg) [reply presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: Added a line to address this.

File: presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+particular, broadcast joins will be faster if the build side is much smaller
+than the probe side. However, broadcast joins require that the tables on the
+build side of the join after filtering fit in memory on each node, whereas
+distributed joins only need to fit in distributed memory across all nodes.
+
+With cost based join distribution selection, Presto automatically chooses whether to
+use a partitioned or broadcast join. With both cost based join enumeration and cost based join distribution, Presto
+automatically chooses which side is the probe and which is the build.
+
+providing the default value.
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/optimizer/cost-based-optimizations.rst]: I think you accidentally deleted the rest of this paragraph.  It should be something like 

```
The join distribution type is governed by the ``join_distribution_type``
session property, with the ``join-distribution-type`` configuration
property providing the default value.
```

**Comment by rschlussel** [comment]
test failure is a timeout, and this is a pure docs change that's definitely not related.  i'll merge without rebuilding.
